File: learning.py
"""Trial-based associative learning hypothesis constrained by MaleCNS edge support.

Synthetic PN odors, imposed sparsification, reward error, PAM gate, synaptic
plasticity and MBON value readout are engineering assumptions, not recovered biology.
This module does not use the walking model to make choices.
"""
import hashlib
import json
import math
from pathlib import Path
import numpy as np
from scipy.sparse import csr_matrix
from model import BASE

CONDITIONS=['intact','no_plasticity','silence_kc','no_reward','no_dopamine','shuffled']
DEFAULT=dict(seed=7,train_trials=80,probe_trials=40,learning_rate=.2,temperature=.2,
             retention=.995,delay=30,condition='intact',rewarded_odor='A')

class LearningCircuit:
    def __init__(self,path=None):
        self.data=json.loads(Path(path or BASE/'data/learning-circuit.json').read_text())
        self.groups={r:[n for n in self.data['nodes'] if n['role']==r] for r in ['PN','KC','MBON','PAM']}
        self.maps={r:{n['bodyId']:i for i,n in enumerate(ns)} for r,ns in self.groups.items()}
        self.nk=len(self.groups['KC']);self.np=len(self.groups['PN'])
        pn=[];km=[];gate=np.zeros(self.nk)
        for pre,post,w in self.data['edges']:
            if pre in self.maps['PN'] and post in self.maps['KC']: pn.append([self.maps['PN'][pre],self.maps['KC'][post],w])
            elif pre in self.maps['KC'] and post in self.maps['MBON']: km.append([self.maps['KC'][pre],self.maps['MBON'][post],w,pre,post])
            elif pre in self.maps['PAM'] and post in self.maps['KC']:gate[self.maps['KC'][post]]+=w
        self.pn=np.array(pn);self.km=np.array(km)
        self.gate=np.log1p(gate);self.gate/=max(self.gate.max(),1)
        self.base=np.log1p(self.km[:,2]);self.kidx=self.km[:,0].astype(int)

    def encoder(self,condition,seed):
        pre,post,w=self.pn.T
        if condition=='shuffled':post=np.random.default_rng(seed+55000).permutation(post)
        weights=np.log1p(w);norm=np.bincount(post,weights=weights,minlength=self.nk)
        return csr_matrix((weights/np.maximum(norm[post],1),(post,pre)),shape=(self.nk,self.np))

    def run(self,**options):
        p=DEFAULT|options
        if p['condition'] not in CONDITIONS:raise ValueError('Unknown learning condition')
        if p['rewarded_odor'] not in ['A','B']:raise ValueError('Rewarded odor must be A or B')
        for k,lo,hi in [('seed',0,999999),('train_trials',10,200),('probe_trials',10,100),('delay',0,200)]:
            v=float(p[k])
            if not math.isfinite(v) or v!=int(v) or not lo<=v<=hi:raise ValueError(f'{k} must be an integer in [{lo},{hi}]')
            p[k]=int(v)
        for k,lo,hi in [('learning_rate',0,1),('temperature',.05,2),('retention',0,1)]:
            p[k]=float(p[k])
            if not math.isfinite(p[k]) or not lo<=p[k]<=hi:raise ValueError(f'Invalid {k}')
        rng=np.random.default_rng(p['seed'])
        stimulus_rng=np.random.default_rng(p['seed']+1000)
        w=self.encoder(p['condition'],p['seed'])
        # Two abstract odors each activate a random 20% of measured PN cells.
        patterns=np.zeros((2,self.np))
        for i in range(2):patterns[i,stimulus_rng.choice(self.np,max(1,self.np//5),replace=False)]=1
        prototypes=[]
        # Independent noisy test exemplars; no reward information reaches this encoding.
        features={}
        for phase in ['training','testing']:
            examples=[]
            for sample in range(4):
                odors=[]
                for i in range(2):
                    pn=np.clip(patterns[i]+stimulus_rng.normal(0,.035,self.np),0,1)
                    activity=w.dot(pn)
                    threshold=np.quantile(activity,.95)
                    activity=np.maximum(activity-threshold,0)
                    if p['condition']=='silence_kc':activity[:]=0
                    mass=activity[self.kidx]*self.base
                    idx=np.flatnonzero(mass>0)
                    phi=mass[idx]/max(mass.sum(),1e-12)
                    odors.append((idx,phi))
                    if sample==0 and phase=='training':prototypes.append(set(np.flatnonzero(activity)))
                examples.append(odors)
            features[phase]=examples
        delta=np.zeros(len(self.km))
        trials=[];summaries=[];checkpoints=[]
        def digest():return hashlib.sha256(delta.tobytes()).hexdigest()
        rewarded=0 if p['rewarded_odor']=='A' else 1
        schedule=[('Baseline',p['probe_trials'],False,rewarded),('Acquisition',p['train_trials'],True,rewarded),
                  ('Unrewarded probe',p['probe_trials'],False,rewarded),('Delayed memory',p['probe_trials'],False,rewarded),
                  ('Reversal',p['train_trials'],True,1-rewarded),('Reversal probe',p['probe_trials'],False,1-rewarded)]
        for phase,n,training,target in schedule:
            if phase=='Delayed memory':delta*=p['retention']**p['delay']
            before=digest()
            # Counterbalanced left/right layout in each phase; side is never passed to the value function.
            sides=np.arange(n)%2;rng.shuffle(sides)
            probabilities=[];choices=[]
            for j in range(n):
                variant=j%4
                fs=features['training' if training else 'testing'][variant]
                values=np.array([np.dot(delta[idx],phi) for idx,phi in fs])
                pa=1/(1+math.exp(float(np.clip((values[1]-values[0])/p['temperature'],-60,60))))
                choice=0 if rng.random()<pa else 1
                delivered=float(training and choice==target and p['condition']!='no_reward')
                error=delivered-float(values[choice]) if training else None
                change=0.
                if training and p['condition'] not in ['no_plasticity','no_dopamine']:
                    idx,phi=fs[choice]
                    gate=self.gate[self.kidx[idx]]
                    denom=float(np.sum(phi*phi*gate))
                    if denom>0:
                        old=delta[idx].copy()
                        delta[idx]=np.clip(old+p['learning_rate']*error*phi*gate/denom,-.95,3)
                        change=float(np.linalg.norm(delta[idx]-old))
                probability_target=pa if target==0 else 1-pa
                trials.append({'trial':len(trials)+1,'phase':phase,'phase_trial':j+1,'training':training,
                               'odor_a_side':'left' if sides[j]==0 else 'right','choice':'AB'[choice],
                               'choice_side':('left' if sides[j]==choice else 'right'),
                               'target_odor':'AB'[target], 'reward':delivered,'p_a':pa,'p_target':probability_target,
                               'value_a':float(values[0]),'value_b':float(values[1]),'prediction_error':error,
                               'weight_change_l2':change,'exemplar':variant})
                probabilities.append(probability_target);choices.append(choice==target)
            after=digest()
            successes=int(sum(choices));z=1.96;prop=successes/n;den=1+z*z/n
            center=(prop+z*z/(2*n))/den;half=z*math.sqrt(prop*(1-prop)/n+z*z/(4*n*n))/den
            summaries.append({'phase':phase,'trials':n,'target_odor':'AB'[target],'choice_fraction':prop,
                              'mean_probability':float(np.mean(probabilities)),'ci95':[center-half,center+half],
                              'reward_count':sum(t['reward'] for t in trials[-n:]),'training':training})
            checkpoints.append({'phase':phase,'before_sha256':before,'after_sha256':after,'weights_changed':before!=after})
        top=np.argsort(np.abs(delta))[-20:][::-1]
        return {'parameters':p,'trials':trials,'summary':summaries,'checkpoints':checkpoints,
                'circuit':{'dataset':self.data['dataset'],'nodes':len(self.data['nodes']),'edges':len(self.data['edges']),
                           'groups':{r:len(n) for r,n in self.groups.items()},'plastic_edges':len(delta),
                           'active_kc_overlap':len(prototypes[0]&prototypes[1])/max(1,len(prototypes[0]|prototypes[1]))},
                'weight_change':{'nonzero_edges':int(np.count_nonzero(delta)),'l2':float(np.linalg.norm(delta)),
                                 'min_multiplier':float((1+delta).min()),'max_multiplier':float((1+delta).max()),
                                 'largest_changes':[{'pre':int(self.km[i,3]),'post':int(self.km[i,4]),'multiplier':float(1+delta[i])} for i in top if delta[i]!=0]},
                'provenance':{'source_files':self.data['files'],'selection':self.data['selection']},
                'interpretation':'Constructed associative learner, not an intelligence score or biological validation. Trial-level rate model is separate from walking. Odor encoding, PAM gating, prediction-error learning, decay and action readout are supplied assumptions. Test phases never update weights; delayed memory applies an explicit decay before testing. Shuffled circuits may also learn.'}

    def suite(self,seed=7):
        rows=[]
        for s in range(int(seed),int(seed)+5):
            for condition in CONDITIONS:
                r=self.run(seed=s,condition=condition,rewarded_odor='A' if s%2 else 'B')
                rows.append({'parameters':r['parameters'],'summary':r['summary']})
        return {'runs':rows,'seeds':5,'conditions':CONDITIONS,'interpretation':'Five initialization/stimulus seeds; original reward A/B counterbalanced across seeds. These are model trials, not biological replicates. Defaults only; no claims of connectome superiority.'}

if __name__=='__main__':
    import argparse
    parser=argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--benchmark',action='store_true')
    parser.add_argument('--seed',type=int,default=7)
    args=parser.parse_args()
    circuit=LearningCircuit()
    result=circuit.suite(args.seed) if args.benchmark else circuit.run(seed=args.seed)
    target=BASE/'data'/('learning-benchmark.json' if args.benchmark else 'learning-example.json')
    target.write_text(json.dumps(result,indent=2))
    print(target)

File: learning.js
'use strict';
const el=id=>document.getElementById('lc-'+id);
const labels={intact:'Intact circuit',no_plasticity:'No plasticity',silence_kc:'KC silenced',no_reward:'No rewards',no_dopamine:'No PAM gate',shuffled:'Shuffled PN → KC'};
let result=null,data=null,trialIndex=0,playing=false,last=0;
const palette={text:'#e7eee5',muted:'#99aa9c',a:'#c2e899',b:'#6fc7bc',reward:'#e9be75',line:'#2b372e'};
for(const [key,n] of [['rate',2],['temp',2],['retain',3]])el(key).oninput=()=>el(key+'-out').textContent=Number(el(key).value).toFixed(n);
const percent=v=>(v*100).toFixed(1)+'%';
function params(){return {condition:el('condition').value,rewarded_odor:el('odor').value,seed:Number(el('seed').value),train_trials:Number(el('train').value),probe_trials:Number(el('probe').value),learning_rate:Number(el('rate').value),temperature:Number(el('temp').value),retention:Number(el('retain').value),delay:Number(el('delay').value)};}
function stroke(ctx,points,color,width=1){ctx.beginPath();points.forEach((p,i)=>i?ctx.lineTo(...p):ctx.moveTo(...p));ctx.strokeStyle=color;ctx.lineWidth=width;ctx.stroke();}
function arena(){
 const c=el('arena'),g=c.getContext('2d'),w=c.width,h=c.height;g.clearRect(0,0,w,h);
 const r=result?.trials[trialIndex];g.font='15px sans-serif';g.fillStyle=palette.muted;
 if(!r){g.fillText('Run the experiment to replay odor choices.',40,60);return;}
 el('phase').textContent=r.phase;el('trial').textContent=`Trial ${r.trial} / ${result.trials.length}`;
 const aLeft=r.odor_a_side==='left',chosenLeft=r.choice_side==='left';
 stroke(g,[[500,385],[500,245],[260,145]],palette.line,68);stroke(g,[[500,245],[740,145]],palette.line,68);
 stroke(g,[[500,385],[500,245],[chosenLeft?260:740,145]],r.choice==='A'?palette.a:palette.b,3);
 for(const [x,odor,side] of [[260,aLeft?'A':'B','left'],[740,aLeft?'B':'A','right']]){
  g.beginPath();g.arc(x,120,60,0,Math.PI*2);g.fillStyle='#18221b';g.fill();g.strokeStyle=odor==='A'?palette.a:palette.b;g.lineWidth=2;g.stroke();
  g.fillStyle=odor==='A'?palette.a:palette.b;g.font='24px Georgia';g.textAlign='center';g.fillText('Odor '+odor,x,126);
  g.font='13px sans-serif';g.fillStyle=palette.muted;g.fillText(side,x,208);
 }
 const x=chosenLeft?325:675,y=174;g.save();g.translate(x,y);g.rotate(chosenLeft?-.95:.95);
 g.fillStyle=palette.text;g.beginPath();g.ellipse(0,0,9,21,0,0,Math.PI*2);g.fill();g.beginPath();g.arc(0,-24,7,0,Math.PI*2);g.fill();
 for(let i=-1;i<=1;i++){stroke(g,[[0,i*10],[-18,i*15+5]],palette.a,2);stroke(g,[[0,i*10],[18,i*15+5]],palette.a,2);}g.restore();
 g.textAlign='left';g.font='15px sans-serif';g.fillStyle=palette.text;g.fillText(`Choice: odor ${r.choice} · ${r.choice_side}`,35,35);
 g.textAlign='right';g.fillStyle=r.reward?palette.reward:palette.muted;g.fillText(r.training?(r.reward?'Sugar reward delivered':'No reward on this choice'):'Unrewarded test · learning frozen',w-35,35);
 g.textAlign='center';g.fillStyle=palette.muted;g.font='14px sans-serif';g.fillText(`P(A) ${percent(r.p_a)}    ·    P(B) ${percent(1-r.p_a)}`,500,430);
 g.textAlign='left';g.font='12px sans-serif';g.fillText('Locations change between trials. The path illustrates the sampled choice.',35,h-13);
 for(const b of el('phases').children)b.classList.toggle('selected',b.textContent===r.phase);
}
function chart(){
 const c=el('chart'),g=c.getContext('2d'),W=c.width,H=c.height;g.clearRect(0,0,W,H);if(!result)return;
 const x=t=>60+(t-1)/(result.trials.length-1)*(W-85),y=p=>H-50-p*(H-82);
 g.font='12px monospace';g.fillStyle=palette.muted;
 for(let i=0;i<=4;i++){const p=i/4;stroke(g,[[60,y(p)],[W-25,y(p)]],palette.line);g.fillText(String(i*25),22,y(p)+4);}
 let offset=0;
 result.summary.forEach((s,i)=>{if(s.training){g.fillStyle='#c2e89908';g.fillRect(x(offset+1),25,x(offset+s.trials)-x(offset+1),H-75);}if(offset)stroke(g,[[x(offset+1),25],[x(offset+1),H-50]],'#75857955');g.fillStyle=palette.muted;g.font='11px sans-serif';g.fillText(s.phase,x(offset+1)+5,16);offset+=s.trials;});
 stroke(g,result.trials.map(r=>[x(r.trial),y(r.p_a)]),palette.a,2.5);
 for(const r of result.trials){g.fillStyle='#e9be7566';g.beginPath();g.arc(x(r.trial),y(r.choice==='A'?1:0),2,0,Math.PI*2);g.fill();}
 stroke(g,[[x(trialIndex+1),25],[x(trialIndex+1),H-50]],palette.text,1);
 g.fillStyle=palette.muted;g.font='12px monospace';for(let i=0;i<=4;i++){const t=Math.round(1+i*(result.trials.length-1)/4);g.fillText(String(t),x(t)-8,H-28);}g.fillText('Trial number',W/2-35,H-7);g.fillText('P(A) %',7,15);
}
function populate(r){
 result=r;trialIndex=0;playing=true;last=0;el('scrub').max=r.trials.length-1;el('scrub').value=0;el('scrub').disabled=false;el('play').disabled=false;el('play').textContent='Pause';el('export').disabled=false;
 for(const [id,index] of [['before',0],['learned',2],['memory',3],['reverse',5]])el(id).textContent=percent(r.summary[index].mean_probability);
 const table=el('summary');table.replaceChildren();r.summary.forEach((s,i)=>{const row=document.createElement('tr');for(const val of [s.phase,s.target_odor,percent(s.mean_probability),`${Math.round(s.choice_fraction*s.trials)} / ${s.trials}`,s.reward_count,r.checkpoints[i].weights_changed?'Updated':'Frozen']){const td=document.createElement('td');td.textContent=String(val);row.append(td);}table.append(row);});
 el('finding').textContent=`${labels[r.parameters.condition]}: preference for the original rewarded odor changes from ${percent(r.summary[0].mean_probability)} to ${percent(r.summary[2].mean_probability)}. After reversal, preference for the newly rewarded odor is ${percent(r.summary[5].mean_probability)}. These are model probabilities; actual sampled choices are shown below.`;
 el('phases').replaceChildren();let offset=0;for(const s of r.summary){const start=offset,b=document.createElement('button');b.textContent=s.phase;b.onclick=()=>{trialIndex=start;playing=false;el('play').textContent='Play';el('scrub').value=start;arena();chart();};el('phases').append(b);offset+=s.trials;}arena();chart();
}
el('run').onclick=async()=>{el('run').disabled=true;playing=false;el('status').textContent='Running acquisition, probes, delay and reversal…';try{const response=await fetch('/api/learn',{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify(params())});const r=await response.json();if(!response.ok)throw Error(r.error);populate(r);el('status').textContent='Complete. Select a phase or compare an ablation.';}catch(e){el('status').textContent=e.message;}finally{el('run').disabled=false;}};
el('play').onclick=()=>{if(trialIndex===result.trials.length-1)trialIndex=0;playing=!playing;last=0;el('play').textContent=playing?'Pause':'Play';};
el('scrub').oninput=()=>{playing=false;trialIndex=Number(el('scrub').value);el('play').textContent='Play';arena();chart();};
el('export').onclick=()=>{const a=document.createElement('a');a.href=URL.createObjectURL(new Blob([JSON.stringify(result,null,2)],{type:'application/json'}));a.download=`odor-learning-${result.parameters.condition}-seed${result.parameters.seed}.json`;a.click();setTimeout(()=>URL.revokeObjectURL(a.href),1000);};
function tick(now){if(playing&&result){if(!last)last=now;if(now-last>220){trialIndex=Math.min(trialIndex+1,result.trials.length-1);last=now;el('scrub').value=trialIndex;arena();chart();if(trialIndex===result.trials.length-1){playing=false;el('play').textContent='Replay';}}}requestAnimationFrame(tick);}
async function init(){arena();try{const response=await fetch('/api/learning-circuit');if(!response.ok)throw Error('Could not load learning circuit');data=await response.json();el('counts').textContent=`${data.nodes.length.toLocaleString()} cells · ${data.edges.length.toLocaleString()} measured edges`;const groups={};for(const n of data.nodes)groups[n.role]=(groups[n.role]||0)+1;el('structure').textContent=`${groups.PN} projection neurons, ${groups.KC.toLocaleString()} Kenyon cells, ${groups.MBON} output neurons and ${groups.PAM} PAM dopamine neurons. Anatomical synapse counts constrain edge support and initial weight scaling.`;el('run').click();
 const br=await fetch('/api/learning-benchmark');if(!br.ok)throw Error('Saved benchmark unavailable');const bench=await br.json();for(const condition of bench.conditions){const runs=bench.runs.filter(r=>r.parameters.condition===condition);const tr=document.createElement('tr');const vals=[labels[condition],...[2,3,5].map(index=>percent(runs.reduce((s,r)=>s+r.summary[index].mean_probability,0)/runs.length))];for(const v of vals){const td=document.createElement('td');td.textContent=v;tr.append(td);}el('benchmark').append(tr);}}
 catch(e){el('status').textContent=e.message;}}
init();requestAnimationFrame(tick);
